From cac45401c75ee9ea9d54689afb54288ca1325773 Mon Sep 17 00:00:00 2001
From: v-zhangxiaomeng5 <v-zhangxiaomeng5@xiaomi.com>
Date: Tue, 12 May 2026 20:10:00 +0800
Subject: [PATCH] libcxx: prioritize include_next wchar.h over glibc
 mbstate_t.h

On NuttX sim platform, only -nostdinc++ is used (not -nostdinc), so the
host glibc headers are visible. This causes a conflict between NuttX's
mbstate_t definition and glibc's bits/types/mbstate_t.h.

Fix by moving the #include_next <wchar.h> branch before the
<bits/types/mbstate_t.h> branch, so NuttX's wchar.h is found first.
---
 include/__mbstate_t.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/__mbstate_t.h b/include/__mbstate_t.h
index c23ea71..18c6dc5 100644
--- a/include/__mbstate_t.h
+++ b/include/__mbstate_t.h
@@ -39,12 +39,12 @@
 #  define __NEED_mbstate_t
 #  include <bits/alltypes.h>
 #  undef __NEED_mbstate_t
+#elif __has_include_next(<wchar.h>)
+#  include_next <wchar.h> // use the C standard provider of mbstate_t if present
 #elif __has_include(<bits/types/mbstate_t.h>)
 #  include <bits/types/mbstate_t.h> // works on most Unixes
 #elif __has_include(<sys/_types/_mbstate_t.h>)
 #  include <sys/_types/_mbstate_t.h> // works on Darwin
-#elif __has_include_next(<wchar.h>)
-#  include_next <wchar.h> // use the C standard provider of mbstate_t if present
 #elif __has_include_next(<uchar.h>)
 #  include_next <uchar.h> // Try <uchar.h> in absence of <wchar.h> for mbstate_t
 #else
-- 
2.34.1

